src/main/java/com/conveyal/r5/analyst/NetworkPreloader.java

```diff
@@ -0,0 +1,241 @@
+import com.conveyal.r5.transit.TransportNetwork;
+import com.conveyal.r5.transit.TransportNetworkCache;
+import com.conveyal.r5.util.AsyncLoader;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.EnumSet;
+import java.util.Objects;
+
+/**
+ * This either returns a network that can be immediately used for a certain analysis task (if it's already prepared)
+ * or kicks off an asynchronous task to prepare such a network, and reports status to the caller. See #381.
+ *
+ * The steps that need to happen to prepare a network for analysis are the following:
+ * 1. Download and load the network if it's already built.
+ * 2. Build the network if it's not already built.
+ * 3. Apply the scenario to the network, or fetch the resulting network if the scenario is already applied.
+ * 3. Pre-compute stop-to-vertex distance tables and linkages (stop-to-point distance tables) for the walking mode.
+ * 4. Create or fetch a grid of destinations based on this specific task and the gridCache.
+ * 5. Link that grid to the street network with the right mode, basing distances on the stop-to-vertex tables created above.
+ *
+ * The network building is done as part of requesting the network.
+ * The distance tables are constructed as part of the scenario application. So the main steps to trigger explicitly are:
+ * 1. Fetch the network
+ * 2. Apply the scenario
+ * 3. Build the linkages
+ *
+ * Implementation notes:
+ *
+ * We don't want to just make all the caches (network, grid, linkage, etc.) asynchronous because we don't want the first
+ * HTTP request to only asynchronously load the network, then the second one to apply the scenario and trigger the
+ * distance table computation, then the third one to link the destination grids for a particular mode. We want all the
+ * steps to be done at one go. Therefore we want a single asynchronous action at the top level, which cascades a lot of
+ * sequential synchronous actions interspersed with progress updates.
+ *
+ * The incoming request may not have a scenario ID in it, in which case it should contain the entire scenario.
+ * TransportNetwork.getNetworkForScenario() resolves it as follows:
+ * String scenarioId = request.scenarioId != null ? request.scenarioId : request.scenario.id;
+ *
+ * I was initially trying to implement this with a Caffeine AsyncLoadingCache, but ran into the following problems:
+ * 1. The cache values must be computed from the key only. But expanded scenarios may be contained in the request.
+ * 2. We want to provide progress information, and retrofitting the Caffeine cache with it would be just as complicated
+ *    as making our own new class.
+ * 3. We don't really want to trigger callbacks after the asynchronous loading is finished, we just want to know
+ *    immediately at the time of each HTTP request whether it's finished building or not.
+ *
+ * Repeatedly re-sending a full Analysis task over HTTP just to see if the network is ready (and failing if it's not)
```

Comment:
Issuing a separate request to prep the network before firing off the actual analysis request seems in line with our long-term goal of explicitly triggering slow operations.

However in this case, the information needed to prep the network (build / load / link / build distance tables) is several fields in the request itself (network and scenario ID, modes of travel, destination grid dimensions). We'd probably end up sending the same request to the prep-network endpoint as to the analysis endpoint itself.

Consider that once the network, linkages, and distance tables are prepared, they may be evicted from cache on the worker. I'm not saying that's ideal, that's just how it works now. So once you've hit the prep-network endpoint and it says "ready", you may hit the analysis endpoint and learn that it's not ready. Doing it with a single endpoint, once the data are ready they are used immediately for analysis. Having separate endpoints means that you always need to hit two endpoints - first to prep the network, then to analyze - for every single point request, and you may need to repeat the cycle if the data are no longer available when you hit the analysis endpoint. It would generate more traffic and not provide any immediate advantage.

Longer term what we want is for the single point analysis endpoint to do only one thing: deserialize the network, linkage, and distance tables from disk. That would be quite fast, on the order of a few seconds. Those serialized, saved data tables would be generated in a separate step. 

So as a development strategy, I'd like to keep the current system (send the task directly to the single-point endpoint without first hitting another endpoint to prime the data), but completely factor out all the graph build, linkage, and distance table calculations into a manually triggered build step that would have its own endpoint. That would have to be done later though.